src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```
@@ -0,0 +1,506 @@
+        AcknowledgeRequest.newBuilder()
+          .setSubscription(subscription)
+          .addAllAckIds(ackIds)
+          .build();
+      subscriber.acknowledgeCallable().call(acknowledgeRequest);
+    }
+  }
+
+  /**
+   * Get Subscriber settings instance.
+   *
+   * @return the Subscriber Stub settings needed to subscribe to a Pub/Sub topic.
+   */
+  protected SubscriberStubSettings getSubscriberSettings() throws IOException {
+    SubscriberStubSettings.Builder builder = SubscriberStubSettings.newBuilder();
+
+    if (credentials != null) {
+      builder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+    }
+
+    return builder.build();
+  }
+
+  /**
+   * Get Subscription Admin Client instance
+   *
+   * @return Subscription Admin Client instance.
+   * @throws IOException
+   */
+  protected SubscriptionAdminClient getSubscriptionAdminClient() throws IOException {
+    return SubscriptionAdminClient.create();
+  }
+
+  /**
+   * Get Subscriber Stub instance
+   *
+   * @param subscriberStubSettings The settings for this Subscriber Stub client.
+   * @return a new Subscriber Stub Instance
+   * @throws IOException the exception thrown by the Pub/Sub library if the Subscriber Stub Settings are invalid.
+   */
+  protected SubscriberStub getSubscriber(SubscriberStubSettings subscriberStubSettings) throws IOException {
+    return GrpcSubscriberStub.create(subscriberStubSettings);
+  }
+
+  /**
+   * Sleep for a given number of milliseconds, calculate new backoff time and return.
+   *
+   * @param backoff the time in milliseconds to delay execution.
+   * @return the new backoff delay in milliseconds
+   */
+  protected int sleepAndIncreaseBackoff(int backoff) {
+    try {
+      if (!isStopped()) {
+        LOG.debug("Backoff - Sleeping for {} ms.", backoff);
+        Thread.sleep(backoff);
+      }
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted Exception in PubSubReceiver.");
```

seems more appropriate to propagate this exception up rather than swallow it here, so that the caller can abort what they're doing. For example, when this is called when creating the subscription, we'll want to stop retrying immediately rather than update the backoff.

---

still have this concern